File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res=nums[0];
        int n=nums.size();
        int prod1=nums[0], prod2=nums[0];
        for(int i=1;i<n;i++){
            if(nums[i]<0)   swap(prod1, prod2);
            prod1 = max(nums[i], prod1*nums[i]);
            prod2 = min(nums[i], prod2*nums[i]);
            res=max(res, prod1);
        }
        return res;
    }
};

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        res = nums[0]
        prod1 = nums[0]
        prod2 = nums[0]

        for i in range(1, n):
            temp = max({nums[i], prod1 * nums[i], prod2 * nums[i]})
            prod2 = min({nums[i], prod1 * nums[i], prod2 * nums[i]})
            prod1 = temp
            res = max(res, prod1)
        return res